### PR TITLE
Using remote LibreOffice Online server on demand

### PR DESCRIPTION
**Pull request by Wastack**

### CODE REVIEW

**Comment by sbraconnier** [comment]
Thanks for this pull request. It seems to be in conflict with some recent changes I made. I'll take a look and come back to you.

File: jodconverter-cli/src/integTest/java/org/jodconverter/cli/ConvertITest.java [hunk 1]
```diff
@@ -187,6 +188,43 @@ public void convert_WithSingleFilter_ShouldSucceed() throws Exception {
   }
 
   @Test
+  public void main_Convert_WithConnectionOption_WithOutputFormat_ShouldSucceed() throws Exception {
+    final File inputFile = new File(SOURCE_FILE);
+    final File outputFile =
+        new File(
+            inputFile.getParentFile(), FilenameUtils.getBaseName(inputFile.getName()) + ".pdf");
+
+    assertThat(outputFile).doesNotExist();
```
Comment (javydreamercsw) [inline jodconverter-cli/src/integTest/java/org/jodconverter/cli/ConvertITest.java]: If for some reason the test fails after creating the file but before deleting it without throwing an exception the file is not deleted so this assertion will fail. See line 223.

File: jodconverter-core/build.gradle [hunk 2]
```diff
@@ -10,6 +10,11 @@ dependencies {
   compile libraries.office_unoil
   compile libraries.slf4j_api
 
+  // https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient
```
Comment (javydreamercsw) [inline jodconverter-core/build.gradle]: Commented code should be removed.

File: jodconverter-core/src/main/java/org/jodconverter/task/COnlineConversationTask.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    this.connectionURL = connectionURL;
+    this.defaultLoadProperties = defaultLoadProperties;
+    this.filterChain = filterChain;
+  }
+
+  @Override
+  public void execute(OfficeContext context) throws OfficeException {
+    try (CloseableHttpClient httpClient = HttpClientBuilder.create().build()) {
+
+      // TODO set url / choose URL
```
Comment (javydreamercsw) [inline jodconverter-core/src/main/java/org/jodconverter/task/COnlineConversationTask.java]: Seems like this comment no longer applies.

**Comment by sbraconnier** [comment]
I think I got the idea behind this pull request, and it is a good one.

However, some details bother me:
- I don't really like the fact that this pull request add some dependencies (http-components) to the core module, since most of the time, they won't be required.
- The way I see it, setting load/store properties and adding filters to the conversion is pointless using the online mode, since a COnlineConversationTask does not use the defaultLoadProperties and filterChain fields, so they should not be available in the COnlineConverter:
```java
COnlineConverter converter = COnlineConverter.make(...);
converter.convert(source).filterWith(...).to(target); // The filters would never be applied... This is bad. The filterWith should not be available here.
```
- The current JODConverter implementation requires an OOo installation, and the COnlineConverter does not. It would be pointless to execute online conversions if you actually need to install office (Libre or Open) in order to use the online converter.

I really like the idea, but it needs to be rethink a little.

I would like to create 2 new modules that would depend on the core module; jodconverter-local and jodconverter-online, and put all the classes related to local conversion into the local module and all the new stuff related to online conversion into the online module. This way, the online module won't have all the OOo libraries as dependencies, nor the local module would have the http-components libraries as dependencies.

I would also adjust a little the `job` package so the online converter won't offer to customize load/store properties and filters.

It's not necessary a big deal but it takes time and right now I'm dreadfully busy. I will integrate this pull request into my local environment and I'll probably start working on this at the end of next week. 

Thanks for this work, it will be a big success :thumbsup:

**Comment by javydreamercsw** [comment]
Sounds great @sbraconnier. It's maybe time of considering a plug-in system?

**Comment by Wastack** [comment]
Thank you for your remarks and your help.

Please let me know if there is anything I can help you to make the changes you advised above, or if there are further improvements needed.

**Comment by sbraconnier** [comment]
@javydreamercsw You are probably right! But this would also probably mean a major rewrite of the project. I'm not sure we are there yet. I didn't know about the LibreOffice Online server project (which seems to be still a work in progress), but I doubt there will be another type of converter for Java Open Document conversion.

@Wastack You can help for sure. I'm about to commit a major refactoring. You could take a look at the new jodconverter-online project. Finally, I choose to create a new OfficeManager for online conversions. I felt like the manager should take care of creating the HttpClient, not the task itself. The online module looks like an incomplete (or limited) one for now (it does not support https request, etc). But since I cannot test it using a true LibreOffice Online installation, it is the best we can do for now.

Maybe you could help me find more documentation about LibreOffice Online? It would help to create some wiki pages about online conversions.

**Pull request merged**